File: ai_houkai/installers/opencode.py
"""
OpenCode installer for the AI-Houkai MCP server.

Registers `ai-houkai-mcp` in OpenCode's config so the agent gains a persistent
memory. OpenCode (sst/opencode) uses its own `mcp` schema — distinct from the
`mcpServers` block used by Claude/Cursor: each server is an entry under `mcp`
with `type: "local"`, a `command` *array*, an `environment` object, and an
`enabled` flag.

    ~/.config/opencode/opencode.json   global (all projects)
    <project>/opencode.json            project-scoped

Library use:

    from ai_houkai.installers import OpenCodeInstaller

    inst = OpenCodeInstaller(memory_path="~/.ai_houkai/.chroma")
    inst.install()                # patch ~/.config/opencode/opencode.json
    inst.print_config()           # preview the JSON block
    inst.verify()                 # smoke-test the server
    print(inst.agents_snippet())  # AGENTS.md content

CLI use (also exposed as the `ai-houkai-install-opencode` console script):

    python -m ai_houkai.installers.opencode --install
    python -m ai_houkai.installers.opencode --project       # ./opencode.json
    python -m ai_houkai.installers.opencode --verify
    python -m ai_houkai.installers.opencode --agents

Everything that is not OpenCode-specific — the merge-and-write install, the
preview, verify, and the argparse front end — lives in
:class:`ai_houkai.installers.common.JSONConfigInstaller`.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import ClassVar, Optional

from ai_houkai.installers.common import (
    DEFAULT_MEMORY_PATH,
    MEMORY_GUIDE,
    SERVER_NAME,
    JSONConfigInstaller,
)

GLOBAL_CONFIG_PATH  = os.path.expanduser("~/.config/opencode/opencode.json")
PROJECT_CONFIG_PATH = "opencode.json"
DEFAULT_COLLECTION  = "opencode"
CONFIG_SCHEMA_URL   = "https://opencode.ai/config.json"

# Re-exported from common so `from ai_houkai.installers.opencode import
# DEFAULT_MEMORY_PATH` keeps working; the values are shared across clients.
__all__ = [
    "CONFIG_SCHEMA_URL",
    "AGENTS_SNIPPET",
    "DEFAULT_COLLECTION",
    "DEFAULT_MEMORY_PATH",
    "GLOBAL_CONFIG_PATH",
    "OpenCodeInstaller",
    "PROJECT_CONFIG_PATH",
    "SERVER_NAME",
]


# OpenCode reads project/global instructions from AGENTS.md.
AGENTS_SNIPPET = f"## Memory (AI-Houkai MCP)\n\n{MEMORY_GUIDE}"


@dataclass
class OpenCodeInstaller(JSONConfigInstaller):
    """Register the AI-Houkai MCP server with OpenCode."""

    client_name:         ClassVar[str] = "OpenCode"
    slug:                ClassVar[str] = "opencode"
    config_key:          ClassVar[str] = "mcp"
    default_collection:  ClassVar[str] = DEFAULT_COLLECTION
    global_config_path:  ClassVar[str] = GLOBAL_CONFIG_PATH
    project_config_path: ClassVar[str] = PROJECT_CONFIG_PATH
    preview_hint:        ClassVar[str] = (
        "Then restart OpenCode — '{server_name}' tools become available "
        "to the agent.")
    installed_hint:      ClassVar[str] = "Restart OpenCode to load the memory tools."
    snippet_flag:        ClassVar[str] = "agents"
    snippet_help:        ClassVar[str] = "Print an AGENTS.md memory-usage snippet"
    snippet_heading:     ClassVar[str] = "AGENTS.md snippet"
    snippet:             ClassVar[str] = AGENTS_SNIPPET

    def config_defaults(self) -> dict:
        return {"$schema": CONFIG_SCHEMA_URL}

    def build_mcp_block(self) -> dict:
        return {
            "type":        "local",
            "command":     [self.mcp_command],
            "enabled":     True,
            "environment": self.build_env(),
        }

    @staticmethod
    def agents_snippet() -> str:
        return AGENTS_SNIPPET


def _main(argv: Optional[list] = None) -> int:
    return OpenCodeInstaller.main(argv)


if __name__ == "__main__":
    raise SystemExit(_main())

File: go/internal/installer/opencode.go
package installer

// OpenCode installer for the AI-Houkai MCP server.
//
// Registers `ai-houkai-mcp` in OpenCode's config so the agent gains a
// persistent memory. OpenCode (sst/opencode) uses its own `mcp` schema —
// distinct from the `mcpServers` block used by Claude/Cursor: each server is
// an entry under `mcp` with `type: "local"`, a `command` *array*, an
// `environment` object, and an `enabled` flag.
//
//	~/.config/opencode/opencode.json   global (all projects)
//	<project>/opencode.json            project-scoped

import (
	"fmt"
	"os"
	"path/filepath"
)

// OpenCodeProjectConfigPath is the project-scoped OpenCode config location.
const OpenCodeProjectConfigPath = "opencode.json"

// OpenCodeConfigSchemaURL is the $schema value OpenCode configs carry.
const OpenCodeConfigSchemaURL = "https://opencode.ai/config.json"

// OpenCodeAgentsSnippet is the AGENTS.md content — OpenCode reads
// project/global instructions from AGENTS.md.
const OpenCodeAgentsSnippet = "## Memory (AI-Houkai MCP)\n\n" + MemoryGuide

// OpenCodeInstaller registers the AI-Houkai MCP server with OpenCode.
type OpenCodeInstaller struct {
	MemoryPath   string
	Collection   string
	SettingsPath string
	ServerName   string
	BinaryPath   string // path to ai-houkai-mcp; defaults to ResolveMCPCommand()
	ExtraEnv     map[string]string
}

// DefaultOpenCodeInstaller targets the global ~/.config/opencode/opencode.json.
func DefaultOpenCodeInstaller() OpenCodeInstaller {
	home, _ := os.UserHomeDir()
	return OpenCodeInstaller{
		MemoryPath:   filepath.Join(home, ".ai_houkai"),
		Collection:   "opencode",
		SettingsPath: filepath.Join(home, ".config", "opencode", "opencode.json"),
		ServerName:   "ai-houkai",
	}
}

func (i OpenCodeInstaller) buildMCPBlock() map[string]any {
	return map[string]any{
		"type":        "local",
		"command":     []string{mcpCommand(i.BinaryPath)},
		"enabled":     true,
		"environment": serverEnv(i.MemoryPath, i.Collection, i.ExtraEnv),
	}
}

// Install patches opencode.json with the MCP server block and returns the
// written path.
func (i OpenCodeInstaller) Install() (string, error) {
	return mergeServerBlock(i.SettingsPath, "mcp", i.ServerName,
		i.buildMCPBlock(),
		map[string]any{"$schema": OpenCodeConfigSchemaURL})
}

// Verify returns true if the server entry exists in opencode.json.
func (i OpenCodeInstaller) Verify() bool {
	return hasServer(i.SettingsPath, "mcp", i.ServerName)
}

// PrintConfig prints the MCP block for manual pasting.
func (i OpenCodeInstaller) PrintConfig() {
	printPasteBlock(i.SettingsPath,
		map[string]any{
			"$schema": OpenCodeConfigSchemaURL,
			"mcp":     map[string]any{i.ServerName: i.buildMCPBlock()},
		},
		fmt.Sprintf("Then restart OpenCode — %q tools become available to the agent.", i.ServerName))
}
